fix(common): do not round up fractions of a millisecond in `DatePipe`

ajitsinghkaler opened the pull request:
Date pipe rounds up milliseconds when passed fractional dates.
ECMAScript spec defines that non-integer values passed for any of the parameters to
new Date(year, month, date, hours, minutes, seconds, ms)
should call floor() to remove the fractional part.

BREAKING CHANGE:  Some people may rely on the rounding functionality of date pipe
Fixes #37989

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Date pipe rounds fractional part of seconds

Issue Number: #37989 


## What is the new behavior?
Datepipe floors fractional seconds


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ajitsinghkaler commented:
Test failures seem unrelated

petebacondarwin reviewed (changes requested):
Thanks for putting this PR together @ajitsinghkaler. Can we tweak the commit message a bit.

1. `Fixes` lines should come before `BREAKING CHANGE` blocks. Otherwise they become part of the breaking change notice.
2. We should refer to `DatePipe` rather than `date pipe` since that is the name of the class.
3. BREAKING CHANGE notices should describe what has changed, what type of application code could be affected, and, if possible, how to update your code to cope if you are affected.

My suggestion for the commit message would be:

```
fix(common): do not round up fractions of a millisecond in `DatePipe`

Currently, the `DatePipe` (via `formatDate()` rounds fractions of a millisecond to the
nearest millisecond. This can cause dates that are less than a millisecond before midnight
to be incremented to the following day.

The [ECMAScript specification](https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.11)
defines that `DateTime` milliseconds should always be rounded down, so that `999.9ms`
becomes `999ms`.

This change brings `formatDate()` and so `DatePipe` inline with the ECMAScript
specification.

Fixes #37989

BREAKING CHANGE:

When passing a date-time formatted string to the `DatePipe` in a format that contains
fractions of a millisecond, the milliseconds will now always be rounded down rather than
to the nearest millisecond.

Most applications will not be affected by this change. If this is not the desired behaviour
then consider pre-processing the string to round the millisecond part before passing
it to the `DatePipe`.
```

AndrewKushnir commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -734,7 +734,7 @@ export function isoStringToDate(match: RegExpMatchArray): Date {
   const m = Number(match[5] || 0) - tzMin;
   const s = Number(match[6] || 0);
-  const ms = Math.round(parseFloat('0.' + (match[7] || 0)) * 1000);
+  const ms = Math.floor(parseFloat('0.' + (match[7] || 0)) * 1000);
```
Thanks for the fix @ajitsinghkaler 👍 I think it'd be great to capture the reason why `Math.floor` is used here (as a comment in the code), similar to what you have in the PR description.

petebacondarwin reviewed (approved):
Great! Just a tiny typo in the commit message... a missing closing bracket.

AndrewKushnir commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -734,7 +734,10 @@ export function isoStringToDate(match: RegExpMatchArray): Date {
-  const ms = Math.round(parseFloat('0.' + (match[7] || 0)) * 1000);
+  // ECMAScript spec defines that parameters to new
+  // Date(year, month, date, hours, minutes, seconds, ms)
+  // should use floor() to remove the fractional part.
```
Thanks for adding this comment @ajitsinghkaler.
I'd propose to reuse a part of the commit message (that also contains a link to ECMAScript spec) here:
```
The ECMAScript specification (https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.11)
defines that `DateTime` milliseconds should always be rounded down, so that `999.9ms`
becomes `999ms`.
```

ajitsinghkaler committed:
fix(common): do not round up fractions of a millisecond in `DatePipe`

Currently, the `DatePipe` (via `formatDate()`) rounds fractions of a millisecond to the
nearest millisecond. This can cause dates that are less than a millisecond before midnight
to be incremented to the following day.

The [ECMAScript specification](https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.11)
defines that `DateTime` milliseconds should always be rounded down, so that `999.9ms`
becomes `999ms`.

This change brings `formatDate()` and so `DatePipe` inline with the ECMAScript
specification.

Fixes #37989

BREAKING CHANGE:

When passing a date-time formatted string to the `DatePipe` in a format that contains
fractions of a millisecond, the milliseconds will now always be rounded down rather than
to the nearest millisecond.

Most applications will not be affected by this change. If this is not the desired behaviour
then consider pre-processing the string to round the millisecond part before passing
it to the `DatePipe`.

AndrewKushnir reviewed (approved):
Thanks for updating the comment in the code 👍 

Since this is a breaking change, I've also started a [global presubmit](https://test.corp.google.com/ui#id=OCL:320895453:BASE:320895523:1594617086577:1b603c02) in Google's codebase and will share results as soon as I have them.

FYI, I've also added the "blocked" label for now, since we can merge this PR only when master branch becomes available for the changes related to v11 (since this is a breaking change).

Thank you.

AndrewKushnir commented:
FYI, the presubmit went well, so we should be able to land it in v11. We would need to rebase and run another global presubmit before merging (once master becomes available for v11 changes). Thank you.

petebacondarwin commented:
`master` is now open for breaking changes (to land in 11.0.0). So @ajitsinghkaler - would you like to rebase and we can see if we can land this?

ajitsinghkaler commented:
@petebacondarwin rebased it

AndrewKushnir commented:
[Presubmit](http://test/OCL:330625891:BASE:330625913:1599613177174:321bdcae).
